File: card-slider/src/main/java/com/ramotion/cardslider/DefaultViewUpdater.java
package com.ramotion.cardslider;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;

/**
 * Default implementation of {@link CardSliderLayoutManager.ViewUpdater}
 */
public class DefaultViewUpdater implements CardSliderLayoutManager.ViewUpdater {

    public static final float SCALE_LEFT = 0.65f;
    public static final float SCALE_CENTER = 0.95f;
    public static final float SCALE_RIGHT = 0.8f;
    public static final float SCALE_CENTER_TO_LEFT = SCALE_CENTER - SCALE_LEFT;
    public static final float SCALE_CENTER_TO_RIGHT = SCALE_CENTER - SCALE_RIGHT;

    public static final int Z_CENTER_1 = 12;
    public static final int Z_CENTER_2 = 16;
    public static final int Z_RIGHT = 8;

    private int cardWidth;
    private int activeCardLeft;
    private int activeCardRight;
    private int activeCardCenter;
    private float cardsGap;

    private int transitionEnd;
    private int transitionDistance;
    private float transitionRight2Center;

    private CardSliderLayoutManager lm;

    private View previewView;

    @Override
    public void onLayoutManagerInitialized(@NonNull CardSliderLayoutManager lm) {
        this.lm = lm;

        this.cardWidth = lm.getCardWidth();
        this.activeCardLeft = lm.getActiveCardLeft();
        this.activeCardRight = lm.getActiveCardRight();
        this.activeCardCenter = lm.getActiveCardCenter();
        this.cardsGap = lm.getCardsGap();

        this.transitionEnd = activeCardCenter;
        this.transitionDistance = activeCardRight - transitionEnd;

        final float centerBorder = (cardWidth - cardWidth * SCALE_CENTER) / 2f;
        final float rightBorder = (cardWidth - cardWidth * SCALE_RIGHT) / 2f;
        final float right2centerDistance = (activeCardRight + centerBorder) - (activeCardRight - rightBorder);
        this.transitionRight2Center = right2centerDistance - cardsGap;
    }

    @Override
    public void updateView(@NonNull View view, float position) {
        final float scale;
        final float alpha;
        final float z;
        final float x;

        if (position < 0) {
            final float ratio = (float) lm.getDecoratedLeft(view) / activeCardLeft;
            scale = SCALE_LEFT + SCALE_CENTER_TO_LEFT * ratio;
            alpha = 0.1f + ratio;
            z = Z_CENTER_1 * ratio;
            x = 0;
        } else if (position < 0.5f) {
            scale = SCALE_CENTER;
            alpha = 1;
            z = Z_CENTER_1;
            x = 0;
        } else if (position < 1f) {
            final int viewLeft = lm.getDecoratedLeft(view);
            final float ratio = (float) (viewLeft - activeCardCenter) / (activeCardRight - activeCardCenter);
            scale = SCALE_CENTER - SCALE_CENTER_TO_RIGHT * ratio;
            alpha = 1;
            z = Z_CENTER_2;
            if (Math.abs(transitionRight2Center) < Math.abs(transitionRight2Center * (viewLeft - transitionEnd) / transitionDistance)) {
                x = -transitionRight2Center;
            } else {
                x = -transitionRight2Center * (viewLeft - transitionEnd) / transitionDistance;
            }
        } else {
            scale = SCALE_RIGHT;
            alpha = 1;
            z = Z_RIGHT;

            if (previewView != null) {
                final float prevViewScale;
                final float prevTransition;
                final int prevRight;

                final boolean isFirstRight = lm.getDecoratedRight(previewView) <= activeCardRight;
                if (isFirstRight) {
                    prevViewScale = SCALE_CENTER;
                    prevRight = activeCardRight;
                    prevTransition = 0;
                } else {
                    prevViewScale = ViewCompat.getScaleX(previewView);
                    prevRight = lm.getDecoratedRight(previewView);
                    prevTransition = ViewCompat.getTranslationX(previewView);
                }

                final float prevBorder = (cardWidth - cardWidth * prevViewScale) / 2;
                final float currentBorder = (cardWidth - cardWidth * SCALE_RIGHT) / 2;
                final float distance = (lm.getDecoratedLeft(view) + currentBorder) - (prevRight - prevBorder + prevTransition);

                final float transition = distance - cardsGap;
                x = -transition;
            } else {
                x = 0;
            }
        }

        ViewCompat.setScaleX(view, scale);
        ViewCompat.setScaleY(view, scale);
        ViewCompat.setZ(view, z);
        ViewCompat.setTranslationX(view, x);
        ViewCompat.setAlpha(view, alpha);

        previewView = view;
    }

    protected CardSliderLayoutManager getLayoutManager() {
        return lm;
    }

}

File: card-slider/src/main/java/com/ramotion/cardslider/DefaultViewUpdater.kt
package com.ramotion.cardslider

import android.view.View
import androidx.core.view.ViewCompat
import com.ramotion.cardslider.CardSliderLayoutManager.ViewUpdater

/**
 * Default implementation of [CardSliderLayoutManager.ViewUpdater]
 */
open class DefaultViewUpdater : ViewUpdater {
    private var cardWidth = 0
    private var activeCardLeft = 0
    private var activeCardRight = 0
    private var activeCardCenter = 0
    private var cardsGap = 0f
    private var transitionEnd = 0
    private var transitionDistance = 0
    private var transitionRight2Center = 0f
    protected var layoutManager: CardSliderLayoutManager? = null
        private set
    private var previewView: View? = null
    override fun onLayoutManagerInitialized(lm: CardSliderLayoutManager) {
        layoutManager = lm
        cardWidth = lm.cardWidth
        activeCardLeft = lm.activeCardLeft
        activeCardRight = lm.activeCardRight
        activeCardCenter = lm.activeCardCenter
        cardsGap = lm.cardsGap
        transitionEnd = activeCardCenter
        transitionDistance = activeCardRight - transitionEnd
        val centerBorder = (cardWidth - cardWidth * SCALE_CENTER) / 2f
        val rightBorder = (cardWidth - cardWidth * SCALE_RIGHT) / 2f
        val right2centerDistance = activeCardRight + centerBorder - (activeCardRight - rightBorder)
        transitionRight2Center = right2centerDistance - cardsGap
    }

    override fun updateView(view: View, position: Float) {
        val scale: Float
        val alpha: Float
        val z: Float
        val x: Float
        if (position < 0) {
            val ratio = (layoutManager?.getDecoratedLeft(view)?.toFloat() ?: 0f) / activeCardLeft
            scale = SCALE_LEFT + SCALE_CENTER_TO_LEFT * ratio
            alpha = 0.1f + ratio
            z = Z_CENTER_1 * ratio
            x = 0f
        } else if (position < 0.5f) {
            scale = SCALE_CENTER
            alpha = 1f
            z = Z_CENTER_1.toFloat()
            x = 0f
        } else if (position < 1f) {
            val viewLeft = layoutManager?.getDecoratedLeft(view) ?: 0
            val ratio = (viewLeft - activeCardCenter).toFloat() / (activeCardRight - activeCardCenter)
            scale = SCALE_CENTER - SCALE_CENTER_TO_RIGHT * ratio
            alpha = 1f
            z = Z_CENTER_2.toFloat()
            x = if (Math.abs(transitionRight2Center) < Math.abs(transitionRight2Center * (viewLeft - transitionEnd) / transitionDistance)) {
                -transitionRight2Center
            } else {
                -transitionRight2Center * (viewLeft - transitionEnd) / transitionDistance
            }
        } else {
            scale = SCALE_RIGHT
            alpha = 1f
            z = Z_RIGHT.toFloat()
            if (previewView != null && layoutManager != null) {
                val prevViewScale: Float
                val prevTransition: Float
                val prevRight: Int
                val isFirstRight = layoutManager!!.getDecoratedRight(previewView!!) <= activeCardRight
                if (isFirstRight) {
                    prevViewScale = SCALE_CENTER
                    prevRight = activeCardRight
                    prevTransition = 0f
                } else {
                    prevViewScale = previewView?.scaleX ?: 0f
                    prevRight = layoutManager!!.getDecoratedRight(previewView!!)
                    prevTransition = previewView?.translationX ?: 0f
                }
                val prevBorder = (cardWidth - cardWidth * prevViewScale) / 2
                val currentBorder = (cardWidth - cardWidth * SCALE_RIGHT) / 2
                val distance = layoutManager!!.getDecoratedLeft(view) + currentBorder - (prevRight - prevBorder + prevTransition)
                val transition = distance - cardsGap
                x = -transition
            } else {
                x = 0f
            }
        }
        view.scaleX = scale
        view.scaleY = scale
        ViewCompat.setZ(view, z)
        view.translationX = x
        view.alpha = alpha
        previewView = view
    }

    companion object {
        const val SCALE_LEFT = 0.65f
        const val SCALE_CENTER = 0.95f
        const val SCALE_RIGHT = 0.8f
        const val SCALE_CENTER_TO_LEFT = SCALE_CENTER - SCALE_LEFT
        const val SCALE_CENTER_TO_RIGHT = SCALE_CENTER - SCALE_RIGHT
        const val Z_CENTER_1 = 12
        const val Z_CENTER_2 = 16
        const val Z_RIGHT = 8
    }
}
